src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,290 @@
+  for (auto offset = ChunkOffset{0}; offset < n_values; ++offset) {
+    auto value = dictionary[attribute_vector_ptr->get(offset)];
+    if (matches_search_value<T>(value)) {
+      include_rows_ptr->emplace_back(offset);
+    }
+  }
+  return include_rows_ptr;
+}
+
+template <typename T>
+std::shared_ptr<std::vector<ChunkOffset>> TableScan::scan_segment(
+    const std::shared_ptr<const ReferenceSegment> segment_ptr) const {
+  // determine which values match the filter condition in a
+  // reference segment.
+  // We cannot determine if a row matches the filter condition
+  // based just on the information in the reference segment. We
+  // need to go to the table that the reference segment points to
+  // in order to retrieve the actual values and perform the filtering.
+  // We need to treat the reference segment differently based on whether
+  // it points to a value segment or a dictionary segment.
+
+  auto include_rows_ptr = std::make_shared<std::vector<ChunkOffset>>();
+  auto referenced_table_ptr = segment_ptr->referenced_table();
+  auto referenced_column_id = segment_ptr->referenced_column_id();
+  auto pos_list_ptr = segment_ptr->pos_list();
+  auto n_values = pos_list_ptr->size();
+
+  // iterate over all rows in the reference segment. Retrieve the
+  // segment that the reference segment points to and get the actual
+  // value from that segment.
+  for (auto offset = ChunkOffset{0}; offset < n_values; ++offset) {
+    auto row_id = (*pos_list_ptr)[offset];
+    auto chunk_id = row_id.chunk_id;
+    auto chunk_offset = row_id.chunk_offset;
+    auto referenced_segment_ptr = referenced_table_ptr->get_chunk(chunk_id)->get_segment(referenced_column_id);
+
+    // referenced segment is ValueSegment
+    const auto typed_value_segment_ptr = std::dynamic_pointer_cast<ValueSegment<T>>(referenced_segment_ptr);
+    if (typed_value_segment_ptr) {
+      auto value = (typed_value_segment_ptr->values())[chunk_offset];
+      if (matches_search_value<T>(value)) {
+        include_rows_ptr->emplace_back(offset);
+      }
+      continue;
+    }
+
+    // referenced segment is DictionarySegment
+    const auto typed_dict_segment_ptr = std::dynamic_pointer_cast<DictionarySegment<T>>(referenced_segment_ptr);
+    if (typed_dict_segment_ptr) {
+      auto dictionary = typed_dict_segment_ptr->dictionary();
+      auto attribute_vector_ptr = typed_dict_segment_ptr->attribute_vector();
+      auto value = dictionary[attribute_vector_ptr->get(chunk_offset)];
+      if (matches_search_value<T>(value)) {
+        include_rows_ptr->emplace_back(offset);
+      }
+      continue;
+    }
+
+    // reference segments can only refer to value segments or dict segments
+    throw std::runtime_error("reference segment refers to invalid segment type");
+  }
+
+  return include_rows_ptr;
+}
+
+template <typename T>
+bool TableScan::matches_search_value(T value) const {
+  auto search_value = type_cast<T>(_search_value);
+  switch (_scan_type) {
```

Comment:
You are calling this function inside the hot loop - an alternative would be to resolve a comparator only once outside the loop and reuse it, as it does not change within a scan. E.g., by using `std::equal_to` etc. See: [here](https://en.cppreference.com/w/cpp/utility/functional/equal_to)

Reply:
Good point!